Add rerank and sentence-similarity tasks to text embedding module

markstur opened the pull request:
Add rerank and sentence-similarity tasks to text embedding module.
This PR is stacked on the first embedding PR #224.
Since embedding, sentence-similarity, and rerank all work with a sentence-transformer model this module will load a model and is able to run multiple tasks.

In addition to the 3 (embed, rerank, sentence-similarity), there are another 3 so that each is not limited to a single input string.  In the case of sentence-similarity and rerank this means that a list of source_sentences or queries (respectively) are each applied against the same list of sentences or documents.  So there is a real benefit many queries can be sent against a large collection of documents.  In the case of embeddings, this is simple batching.

Text Embedding Module
=====================

Implements the following tasks:

  1. EmbeddingTask: Returns an embedding from an input text string
  2. EmbeddingsTasks: EmbeddingTask but with a list of inputs producing a list of outputs
  3. SentenceSimilarityTask: Compare one source sentence to a list of sentences
  4. SentenceSimilarityTasks: SentenceSimilarityTask but with a list of source sentences producing
     a list of outputs
  5. RerankTask: Return top_n documents ordered by relevance given a query
  6. RerankTasks: RerankTask but with a list of queries producing a list of outputs

More details for sentence-similarity and rerank...

sentence-similarity is a common and simple concept (see Hugging Face or Sentence Transformers)

class SentenceSimilarityTask(TaskBase):
    """Compare the source_sentence to each of the sentences.
    Result contains a list of scores in the order of the input sentences.
    """

@task(
    required_parameters={"source_sentences": List[str], "sentences": List[str]},
    output_type=SentenceListScores
)
class SentenceSimilarityTasks(TaskBase):
    """Compare each of the source_sentences to each of the sentences.
    Returns a list of results in the order of the source_sentences.
    Each result contains a list of scores in the order of the input sentences.
    """

rerank is less intuitive, but is popular for RAG and chaining.  One of the more popular rerank APIs is from Cohere.  This implementation is similar to their API.

class RerankTask(TaskBase):
    """Returns an ordered list ranking the most relevant documents for the query

    Required parameters:
        query: The search query
        documents: JSON documents containing "text" or alternative "_text" to search
    Returns:
        The top_n documents in order of relevance (most relevant first).
        For each, a score and document index (position in input) is returned.
        The original document JSON is returned depending on optional args.
        The top_n optional parameter limits the results when used.
    """

class RerankTasks(TaskBase):
    """Returns an ordered list for each query ranking the most relevant documents for the query

    Required parameters:
        queries: The search queries
        documents: JSON documents containing "text" or alternative "_text" to search
    Returns:
        Results in order of the queries.
        In each query result:
            The query text is optionally included for visual convenience.
            The top_n documents in order of relevance (most relevant first).
            For each, a score and document index (position in input) is returned.
            The original document JSON is returned depending on optional args.
            The top_n optional parameter limits the results when used.
    """

gkumbhat commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+        if len(queries) < 1 or len(documents) < 1:
+            return RerankPrediction([])
+
+        if top_n is None or top_n < 1:
```
this default behavior is undocumented. Can we please add this in docstrings?

markstur replied:
done

markstur commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+            error.dir_check("<NLP01027299E>", model_name_or_path)
+        else:
+            # If no artifacts_path, look for hf_model Hugging Face model by name (or path)
+            model_name_or_path = config.get(cls._HF_HUB_KEY)
```
has been removed

markstur commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+            top_n = len(documents)
+
+        # Using input document dicts so get "text" else "_text" else default to ""
+        doc_texts = [srd.get("text") or srd.get("_text", "") for srd in documents]
```
The comparison is typically with the "text" field of the JSON document.  We also have cases where "_text" is used instead.  So this impl uses "text" if found, but if not uses "_text" as the alternate.

Currently if neither is found text="" is used.  This almost makes sense, but if that is a real use case we might need to handle it better.

markstur commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+
+        query_embeddings = self.model.encode(queries, convert_to_tensor=True)
+        query_embeddings = query_embeddings.to(self.model.device)
+        query_embeddings = normalize_embeddings(query_embeddings)
```
sentence-transformers automatically does the `if cuda use gpu` logic (with env var controls).  So here we use the device that was set on self.model.device and get all the embeddings to() there as well before scoring.  This is the from the sentence-transformers optimization examples.

Performance eval for whether or not to use GPU here is not done yet, but we at least have environment control (e.g., per pod).

markstur commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+        self.model = model
+
+    @classmethod
+    def load(cls, model_path: str) -> "Rerank":
```
Now that we have multi-task, I've combined them.
The embedding PR is open.  Now I'm stacking this PR on that one.  Wanted to do smaller PRs, but of course there is a dependency and I think for rerank and sentence-similarity it makes sense to see both at once for review/discussion.

markstur commented on caikit_nlp/modules/reranker/rerank_task.py:
```diff
@@ -0,0 +1,39 @@
+    },
+    output_type=RerankPrediction,
+)
+class RerankTask(TaskBase):
```
added

markstur commented on caikit_nlp/data_model/reranker.py:
```diff
@@ -0,0 +1,55 @@
+class RerankScore(DataObjectBase):
+    """The score for one document (one query)"""
+
+    document: JsonDict
```
the desired API is a document that is JSON with a text (or alternative _text) field that is used for ranking while the rest of the document is typically returned reranked.  JsonDict works for me for gRPC and REST while allowing different types input/output even nested.  I'm not sure if you are recommending a preferred alternative.

If you are just wondering why not only pass text and return index, then I understand (agree) but that isn't the requested API for the rerank use case.

markstur commented on caikit_nlp/data_model/reranker.py:
```diff
@@ -0,0 +1,55 @@
+class RerankPrediction(DataObjectBase):
+    """Result for a rerank task"""
+
+    results: List[RerankQueryResult]
```
1 query with the top_n document results in order of relevance for that query.

markstur replied:
done

markstur commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+        model: SentenceTransformer,
+    ):
+        """Initialize
+        This function gets called by `.load` and `.train` function
```
removed.  It's not a good practice to try to document all usages and __init__() doesn't need "Initialize" called out.

markstur commented on caikit_nlp/modules/reranker/rerank.py:
```diff
@@ -0,0 +1,198 @@
+        Args:
+            queries: List[str]
+            documents:  List[JsonDict]
+            top_n:  Optional[int]
```
done and also renamed RerankPrediction -> RerankPredictions because this is used to provide a result for each of the queries (plural).

markstur commented:
Checked feedback again and added more docstring for some things I had missed.  Thanks @gkumbhat

taneem-ibrahim commented:
@dtrifiro @Xaenalt could you please help review this PR? Thank you!

dtrifiro commented on tests/data_model/test_reranker.py:
```diff
@@ -0,0 +1,134 @@
+input_random_score = {
+    "document": input_random_document,
+    "index": random.randint(-99999, 99999),
+    "score": random.uniform(-99999, 99999),
+    "text": "".join(random.choices(string.printable, k=100)),
+}
+
+input_random_score_3 = {
+    "document": {"text": "random foo3"},
+    "index": random.randint(-99999, 99999),
+    "score": random.uniform(-99999, 99999),
+    "text": "".join(random.choices(string.printable, k=100)),
+}
+
+input_scores = [dm.RerankScore(**input_score), dm.RerankScore(**input_random_score)]
+input_scores2 = [
+    dm.RerankScore(**input_random_score),
+    dm.RerankScore(**input_random_score_3),
+]
+
+input_result_1 = {"query": "foo", "scores": input_scores}
+input_result_2 = {"query": "bar", "scores": input_scores2}
+input_results = [
+    dm.RerankQueryResult(**input_result_1),
+    dm.RerankQueryResult(**input_result_2),
+]
+
+input_sentence_similarity_scores_1 = {
+    "scores": [random.uniform(-99999, 99999) for _ in range(10)]
+}
+input_sentence_similarity_scores_2 = {
+    "scores": [random.uniform(-99999, 99999) for _ in range(10)]
+}
+
+input_sentence_similarities_scores = [
+    dm.SentenceScores(**input_sentence_similarity_scores_1),
+    dm.SentenceScores(**input_sentence_similarity_scores_2),
+]
+
+
```
All of these could be pytest fixtures

markstur replied:
done

dtrifiro commented on tests/data_model/test_reranker.py:
```diff
@@ -0,0 +1,134 @@
+    assert new_do_from_init == new_do_from_json
+
+
+def assert_fields_match(data_object, inputs):
+    for k, v in inputs.items():
+        assert getattr(data_object, k) == inputs[k]
```
```suggestion
def assert_fields_match(data_object, inputs):
    assert all(getattr(data_object, key) == value for key, value in inputs.items())
```

markstur replied:
done

dtrifiro reviewed (approved):
Left a few minor comments. Looks great overall. My only real concern is using `sentence-transformers, since it doesn't look like it's being actively maintained

markstur commented on pyproject.toml:
```diff
@@ -24,6 +24,7 @@ dependencies = [
     "pandas>=1.5.0",
     "scikit-learn>=1.1",
     "scipy>=1.8.1",
+    "sentence-transformers~=2.2.2",
```
I think the popularity still goes to sentence-transformers (unless my searches are biased), but the maintenance activity does seem slow.  The workaround is to use transformers feature-extraction directly and add mean pooling and normalization (and cosine, dot_score...).  Hugging Face seems to prefer to defer to sentence-transformers for now as far as I can tell.  Old requests to create HF pipelines to replace sentence-transformers were rejected.

markstur commented on caikit_nlp/data_model/embedding_vectors.py:
```diff
@@ -0,0 +1,199 @@
+        return cls(data=data)
+
+    @classmethod
+    def from_json(cls, json_str):
```
done, but didn't rename the arg.  Want to keep it in sync with the base.

markstur commented on caikit_nlp/data_model/reranker.py:
```diff
@@ -0,0 +1,50 @@
+from caikit.core.data_model.json_dict import JsonDict
+
+
+@dataobject()
```
good catch.  Updated them all to specify this package (fwiw).

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+    @EmbeddingTasks.taskmethod()
+    def run_embeddings(
+        self, texts: List[str]  # pylint: disable=redefined-builtin
+    ) -> ListOfVector1D:
```
another good catch!  Previously was "input" which is a builtin.  So I removed all the obsolete pylint comments.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+    @EmbeddingTask.taskmethod()
+    def run_embedding(
+        self, text: str
+    ) -> EmbeddingResult:  # pylint: disable=redefined-builtin
```
removed

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+    @SentenceSimilarityTasks.taskmethod()
+    def run_sentence_similarities(
+        self, source_sentences: List[str], sentences: List[str]
+    ) -> SentenceListScores:  # pylint: disable=arguments-differ
```
This was already disabled in .pylintrc so I removed the comments.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+            results[0]
+            if len(results) > 0
+            else RerankQueryResult(scores=[], query=query if return_query else None)
+        )
```
return results[0] but otherwise yep/done.  Looks nicer.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+        )
+
+    @RerankTasks.taskmethod()
+    def run_rerank_queries(
```
I know it's bad to use "historical reasons", but we just very recently went from overriding run() to allowing additional task methods.  So I like having these all start with run_

Any other reviewers disagree with me?   I'll probably eventually wish I followed dtrifiro advice here.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+        )
+
+        if len(queries) < 1 or len(documents) < 1:
+            return RerankPredictions([])
```
There's something I just like about empty iterables, but I looked at this a while and you are right.  It does not make sense, so now it raises ValueError.  Thanks!

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+        def get_text(srd):
+            return srd.get("text") or srd.get("_text", "")
+
+        doc_texts = [get_text(srd) for srd in documents]
```
Took part of your advice (srd -> doc) but I kept get_text function because that decision to use text else _text else "" is in 2 places and I do think we will reconsider that default part.  So at least it is only in once place.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+
+        query_embeddings = self.model.encode(queries, convert_to_tensor=True)
+        query_embeddings = query_embeddings.to(self.model.device)
+        query_embeddings = normalize_embeddings(query_embeddings)
```
Yep, done!

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+                if return_documents:
+                    x["document"] = documents[corpus_id]
+                if return_text:
+                    x["text"] = get_text(documents[corpus_id])
```
here is where I kept that so this "logic" is in one place and see next comment.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+                Each of the queries will be compared to the text of each of the documents.
+            documents:  List[JsonDict]
+                Each document is a dict. The text value is used for comparison to the query.
+                If there is no text key, then _text is used and finally default is "".
```
re: the empty fallback
That is kind of TBD. So I was concerned about getting fewer results than expected in some cases and thought "" was best for now, but perhaps forcing these to score=0 would be a future fix pending feedback TBD.

Today, I'm thinking docs w/o "text" or "_text" should be thrown out and you might get zero docs error or you might get one less result than expected.

I'd prefer that decision to go in a separate PR, but open for review discussion.

markstur replied:
Re: a Document type with those optional fields...
Because this rerank API is apparently used to pipe documents in and return them (e.g. for chaining I suppose) we need the Document to also have other arbitrary keys and values.  I'm not sure we can currently do your suggestion with these extra keys/values today for our data objects for gRPC and FastAPI (pydantic).  I don't think so.

The current code is able to deal with these flexible JsonDict objects (just lacks any structural hint about expected keys).

Maybe @prashantgupta24 has advice.

markstur commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,384 @@
+
+        model_config_path = os.path.abspath(
+            model_config_path.strip()
+        )  # No leading/trailing spaces sneaky weirdness
```
done thanks!

markstur commented on caikit_nlp/modules/text_embedding/rerank_task.py:
```diff
@@ -0,0 +1,73 @@
+# Local
+from caikit_nlp.data_model.reranker import RerankPredictions, RerankQueryResult
+
+logger = alog.use_channel("<SMPL_BLK>")
```
was copy from sample "SMPL".  I've removed these copy/paste error ones because they were not needed.

Thanks again for the good eyes

markstur commented on caikit_nlp/modules/text_embedding/rerank_task.py:
```diff
@@ -0,0 +1,73 @@
+
+@task(
+    required_parameters={
+        "documents": List[JsonDict],
```
See above, but short version is I need the flexibility of any-key, any-value and this does that.

markstur commented on tests/data_model/test_embedding_vectors.py:
```diff
@@ -0,0 +1,130 @@
+random_numpy_vector1d_float64 = random_number_generator.random(
+    DUMMY_VECTOR_SHAPE, dtype=np.float64
+)
+random_python_vector1d_float = random_numpy_vector1d_float32.tolist()
```
done

markstur commented:
I think I addressed all the comments in code and/or in reply.

Sorry for the delay.  I thought I was nearly done and then found a bunch that had been collapsed.

Thanks again for the advice!

markstur commented:
@gkumbhat anything else needed here except for rebase on caikit interfaces and PR #224 

I could just combine the PRs here, but both have been under review.

gkumbhat commented on README.md:
```diff
@@ -8,13 +8,18 @@ Caikit-NLP implements concept of "task" from `caikit` framework to define (and c
+| TextClassificationTask  | 1. `SequenceClassification`                    | 1. (Work in progress..)                                                                                                                                                    |
+| TokenClassificationTask | 1. `FilteredSpanClassification`                | 1. (Work in progress..)                                                                                                                                                    |
+| TokenizationTask        | 1. `RegexSentenceSplitter`                     | 1. Demo purposes only                                                                                                                                                      |
+| EmbeddingTask           | 1. `TextEmbedding`                             | 1. text/embedding from a local sentence-transformers model                                                                                                                 
+| EmbeddingTasks          | 1. `TextEmbedding`                             | 1. Same as EmbeddingTask but multiple sentences (texts) as input and corresponding list of outputs.                                                                        
```
Can we combine these? Like we do for prompt tuning?

markstur replied:
updated README.md please take a look

gkumbhat commented on README.md:
```diff
@@ -8,13 +8,18 @@ Caikit-NLP implements concept of "task" from `caikit` framework to define (and c
+| TokenizationTask        | 1. `RegexSentenceSplitter`                     | 1. Demo purposes only                                                                                                                                                      |
+| EmbeddingTask           | 1. `TextEmbedding`                             | 1. text/embedding from a local sentence-transformers model                                                                                                                 
+| EmbeddingTasks          | 1. `TextEmbedding`                             | 1. Same as EmbeddingTask but multiple sentences (texts) as input and corresponding list of outputs.                                                                        
+| SentenceSimilarityTask  | 1. `TextEmbedding`                             | 1. text/sentence-similarity from a local sentence-transformers model (Hugging Face style API returns scores only in order of input sentences)                              |
+| SentenceSimilarityTasks | 1. `TextEmbedding`                             | 1. Same as SentenceSimilarityTask but multiple source_sentences (each to be compared to same list of sentences) as input and corresponding lists of outputs.               |
+| RerankTask              | 1. `TextEmbedding`                             | 1. text/rerank from a local sentence-transformers model (Cohere style API returns top_n scores in order of relevance with index to source and optionally returning inputs) |
+| RerankTasks             | 1. `TextEmbedding`                             | 1. Same as RerankTask but multiple queries as input and corresponding lists of outputs. Same list of documents for all queries.                                            |
```
Same here

markstur replied:
updated README.md please take a look

gkumbhat commented on caikit_nlp/data_model/sentence_similarity.py:
```diff
@@ -0,0 +1,36 @@
+error = error_handler.get(log)
+
+
+@dataobject(package="caikit_data_model.caikit_nlp")
+class SentenceScores(DataObjectBase):
+    scores: List[float]
+
+
+@dataobject(package="caikit_data_model.caikit_nlp")
+class SentenceListScores(DataObjectBase):
+
+    results: List[SentenceScores]
```
Just realized `producer_id` isn't present in these. May be you can add these in your `caikit` interface PR

markstur replied:
Looks like that is only used in text_gen and only on the task output objects.  I'll try following that pattern if that is the preferred thing going forward.

gkumbhat commented on sentence-transformers.nodeps.txt:
```diff
@@ -0,0 +1,7 @@
+# These can be installed with --no-deps.
```
hm, is this to separate out dependencies before we add other mechanisms?

markstur replied:
I did this because there was some concern about dependencies, but I'll admit this probably is not the best way to deal with it.  I'd actually recommend I remove this and add extras handling like caikit in a separate PR.  Is that preferred?  Or is it better to just accept the dependencies.

gkumbhat commented on caikit_nlp/modules/text_embedding/embedding.py:
```diff
@@ -0,0 +1,413 @@
+    def run_rerank_queries(
+        self,
+        queries: List[str],
+        documents: List[JsonDict],
```
is jsondict and dict same? Wondering if the pure python experience for using these functions will be problematic? i.e if I want to use this function in a notebook or something and then I'll need to first convert my documents to JsonDict ?

markstur replied:
Caikit won't allow Dict[str, Any].  I would have to provide data objects to wrap Any and that is exactly what JsonDict does for me. So this only works with JsonDict unless there is some smaller piece where I can use dict and hide part of the JsonDict usage, but if that is even allowed we'd still have JsonDict output so I think that is not helpful.

So in general, this must be the caikit way, but if you think I can back-off part of that and use Python please clarify and I'll do some more testing to see where that breaks down.   I'd love to use plain Python.

gkumbhat reviewed (commented):
left 2 small comments. Other than that it LGTM

gkumbhat reviewed (approved):
LGTM
